### DISCUSSION TITLE
dav_def_max_size parameter

### DISCUSSION THREAD

**Discussion by cfengno1** [block-hczhai/block2-preview#141] [Q&A] [answered]
Dear Huanchen,

I notice that the default value of dav_def_max_size is 50. This parameter is only useful for determining multiple eigenstates, right? If I only care about the ground state, will this parameter be used and affects the computation time?

Best,
Feng

**Comment by hczhai** [comment]
> This parameter is only useful for determining multiple eigenstates, right?

No. It can affect ground state calculations, if the printed ``Ndav`` is above ``dav_def_max_size``.

> If I only care about the ground state, will this parameter be used and affects the computation time?

Reducing this number can save memory cost, but it can potentially increase the required ``Ndav`` when ``Ndav`` is above ``dav_def_max_size``. And the leading DMRG computational cost is proportional to ``Ndav``. The Davidson threshold can also affect ``Ndav``. Having a loose Davidson threshold can make ``Ndav`` smaller, but the accuracy will also be worse.

Computing more than one root (eigenstates) will significantly increase the required ``Ndav``.

**Reply by cfengno1** [reply]
Dear Huanchen,

Thank you for the explanation!

I was searching the literature to understand the deflation in Davidson algorithm, but it always shows up in the context of finding multiple eigenvectors of a matrix. In the standard setup of the algorithm for the lowest eigenstate (e.g. https://gqcg-res.github.io/knowdes/the-davidson-diagonalization-method.html    and   https://github.com/ITensor/ITensor/blob/v3/itensor/iterativesolvers.h#L88), the deflation is not mentioned. I notice that you use the name "_**generalized**_ Davidson algorithm", so I am wondering what is the method you are using? Can you point to a reference?

Best,
Feng

**Reply by hczhai** [reply, marked as answer]
> In the standard setup of the algorithm for the lowest eigenstate ... the deflation is not mentioned.

``block2::IterativeMatrixFunctions::davidson`` implements the Davidson algorithm (for the standard eigenvalue problem "Hc=Ec") described in the link https://gqcg-res.github.io/knowdes/the-davidson-diagonalization-method.html and Davidson's 1975 paper, and the last paragraph in the link "In case the dimension of the subspace matrix is getting too big..." explains the deflation/restart process. ``dav_def_max_size`` is the "predetermined maximum subspace dimension".

> I notice that you use the name "generalized Davidson algorithm"

The ``block2::IterativeMatrixFunctions::davidson_generalized`` implements the Davidson algorithm for the generalized eigenvalue problem "Hc=ESc" where "S" is an overlap matrix. The generalized eigenvalue problem is not related to the current topic.

**Reply by cfengno1** [reply]
Now I understand. Thank you!
